File: wave2_network/packages/shared/src/encryption/aes.ts
import crypto from 'crypto'

export interface EncryptedPayload {
  iv: Buffer
  authTag: Buffer
  ciphertext: Buffer
}

export function generateKey(): Buffer {
  return crypto.randomBytes(32)
}

export function encryptText(text: string, key: Buffer): EncryptedPayload {
  if (key.length !== 32) {
    throw new Error('Key must be 32 bytes')
  }

  const iv = crypto.randomBytes(16)
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv)
  const ciphertext = Buffer.concat([cipher.update(text, 'utf8'), cipher.final()])
  const authTag = cipher.getAuthTag()

  return { iv, authTag, ciphertext }
}

export function packPayload(payload: EncryptedPayload): Buffer {
  if (payload.iv.length !== 16) {
    throw new Error('IV must be 16 bytes')
  }
  if (payload.authTag.length !== 16) {
    throw new Error('Auth tag must be 16 bytes')
  }

  return Buffer.concat([payload.iv, payload.authTag, payload.ciphertext])
}

export function unpackPayload(packed: Buffer): EncryptedPayload {
  if (packed.length < 32) {
    throw new Error('Packed data too short')
  }

  return {
    iv: packed.subarray(0, 16),
    authTag: packed.subarray(16, 32),
    ciphertext: packed.subarray(32),
  }
}

export function decryptBlob(packed: Buffer, key: Buffer): string {
  if (key.length !== 32) {
    throw new Error('Key must be 32 bytes')
  }

  const { iv, authTag, ciphertext } = unpackPayload(packed)
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv)
  decipher.setAuthTag(authTag)
  const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()])

  return decrypted.toString('utf8')
}

export function splitKeyForFHE(key: Buffer): { high: bigint; low: bigint } {
  if (key.length !== 32) {
    throw new Error('Key must be 32 bytes')
  }

  return {
    high: BigInt(`0x${key.subarray(0, 16).toString('hex')}`),
    low: BigInt(`0x${key.subarray(16, 32).toString('hex')}`),
  }
}

export function combineKeyFromFHE(high: bigint, low: bigint): Buffer {
  const highHex = high.toString(16).padStart(32, '0')
  const lowHex = low.toString(16).padStart(32, '0')

  return Buffer.from(highHex + lowHex, 'hex')
}

File: wave2_network/packages/shared-py/blindference_utils/aes.py
from __future__ import annotations

import os

from Crypto.Cipher import AES


def generate_key() -> bytes:
    return os.urandom(32)


def encrypt_text(text: str, key: bytes) -> dict[str, bytes]:
    if len(key) != 32:
        raise ValueError("Key must be 32 bytes")

    iv = os.urandom(16)
    cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
    ciphertext, auth_tag = cipher.encrypt_and_digest(text.encode("utf-8"))

    return {
        "iv": iv,
        "authTag": auth_tag,
        "ciphertext": ciphertext,
    }


def pack_payload(payload: dict[str, bytes]) -> bytes:
    iv = payload["iv"]
    auth_tag = payload["authTag"]
    ciphertext = payload["ciphertext"]

    if len(iv) != 16:
        raise ValueError("IV must be 16 bytes")
    if len(auth_tag) != 16:
        raise ValueError("Auth tag must be 16 bytes")

    return iv + auth_tag + ciphertext


def unpack_payload(packed: bytes) -> dict[str, bytes]:
    if len(packed) < 32:
        raise ValueError("Packed data too short")

    return {
        "iv": packed[:16],
        "authTag": packed[16:32],
        "ciphertext": packed[32:],
    }


def decrypt_blob(packed: bytes, key: bytes) -> str:
    if len(key) != 32:
        raise ValueError("Key must be 32 bytes")

    payload = unpack_payload(packed)
    cipher = AES.new(key, AES.MODE_GCM, nonce=payload["iv"])
    plaintext = cipher.decrypt_and_verify(payload["ciphertext"], payload["authTag"])
    return plaintext.decode("utf-8")


def split_key_for_fhe(key: bytes) -> tuple[int, int]:
    if len(key) != 32:
        raise ValueError("Key must be 32 bytes")

    high = int.from_bytes(key[:16], "big")
    low = int.from_bytes(key[16:], "big")
    return high, low


def combine_key_from_fhe(high: int, low: int) -> bytes:
    return high.to_bytes(16, "big") + low.to_bytes(16, "big")
